src/main/java/io/strimzi/kafka/metrics/PrometheusMetricsReporterConfig.java

```diff
@@ -60,18 +58,64 @@ private Pattern compileAllowlist(List<String> allowlist) {
         return Pattern.compile(joined);
     }
 
+    public String listener() {
+        return listener.toString();
+    }
+
+    static class Listener {
+
+        private static final Pattern PATTERN = Pattern.compile("http://\\[?([0-9a-zA-Z\\-%._:]*)]?:([0-9]+)");
+
+        final String host;
+        final int port;
+
+        public Listener(String host, int port) {
+            this.host = host;
+            this.port = port;
+        }
+
+        public static Listener parseListener(String listener) {
+            Matcher matcher = PATTERN.matcher(listener);
+            if (matcher.matches()) {
+                String host = matcher.group(1);
+                int port = Integer.parseInt(matcher.group(2));
+                return new Listener(host, port);
+            } else {
+                throw new ConfigException(LISTENER_CONFIG, listener, "Listener be of format http://[host]:[port]");
+            }
+        }
+
+        @Override
+         public String toString() {
+            return "http://" + host + ":" + port;
+        }
+
+        @Override
+         public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Listener listener = (Listener) o;
+            return port == listener.port && Objects.equals(host, listener.host);
+        }
+
+        @Override
+         public int hashCode() {
+            return Objects.hash(host, port);
+        }
+    }
+
     @Override
     public String toString() {
         return "PrometheusMetricsReporterConfig{" +
                 "allowlist=" + allowlist +
-                ", port=" + port +
+                ", listener=" + listener +
                 '}';
     }
 
     public synchronized Optional<HTTPServer> startHttpServer() {
         try {
-            HTTPServer httpServer = new HTTPServer(port, true);
-            LOG.info("HTTP server started on port " + port);
+            HTTPServer httpServer = new HTTPServer(listener.host, listener.port, true);
```

Comment:
I'm still not sure if I have implemented this correctly?? Do you think it would be better to run `startHttpServer` in the `init` method instead of the `configure` method?

Reply:
If the hostname is empty, `InetAddress.getAllByName()` which is called by the `InetSocketAddress()` constructor returns the loopback address


From https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html#getAllByName-java.lang.String-
> If the host is null then an InetAddress representing an address of the loopback interface is returned.

But you're right to be concerned by doing a hostname resolution at startup, I had missed that point. Looking at the code, it seems `HttpServer.create()` ends up calling `new ServerImpl()` which calls `socket.bind()` on the `InetSocketAddress`. This throws `SocketException` if the address is unresolved. I've not tried running the code, so it's worth double checking but I think this is why I told Owen to use the `HttpServer` constructor with a host/port.

Reply:
I wonder if we actually want name resolution at early as possible if the user provided a hostname. Until the HTTP server is started, we won't export metrics and if we can't resolve it we'd rather fail fast than the broker for a moment without exporting metrics.

For reference, this is what jmx_exporter does:
- `premain()`: https://github.com/prometheus/jmx_exporter/blob/main/jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java#L60-L66
- `createHTTPServer()`: https://github.com/prometheus/jmx_exporter/blob/main/jmx_prometheus_common/src/main/java/io/prometheus/jmx/common/http/HTTPServerFactory.java#L104-L115